### PR TITLE
Add in-app support for remote feature flags

### PR DESCRIPTION
**Pull request by jkmassel**
Adds app-level support for remote feature flags.

Should be tested alongside https://github.com/wordpress-mobile/WordPressKit-iOS/pull/273

To test:
1. Apply this PR
2. In the `WordPress` scheme, go to Run > Arguments, and enable `-extra_debug 1`
3. Run the app, and filter logs by the 🚩 character.
4. You should see that the app does the following:
   - Generates a new Device ID (note it for later by copying it to the clipboard or something)
   - Updates the local cache of remote feature flags
5. Close the app
6. Run it again – this time it doesn't generate a new device ID – it uses the same one as before
7. If you'd like to test that the cache works, you can add this code at the top of the AppDelegate's `didFinishLaunching` method:

```swift
debugPrint("🚩 \(remoteFeatureFlagStore.valueForFlag(.unifiedAuth))")
```

It will print `true` to the console.

PR submission checklist:

- [ ] I have considered adding unit tests where possible.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by mindgraffiti** [comment]
cc: @diegoreymendez in case you're curious

Edit: @jkmassel I'm assigning all 3 of us. We have a "1 review then merge it" rule. So once you get an approval, feel free to merge 😁

File: WordPress/Classes/Stores/RemoteFeatureFlagStore.swift
Comment (diegoreymendez) [inline WordPress/Classes/Stores/RemoteFeatureFlagStore.swift]: I'd consider going for a more Swifty:

```swift
public func hasRemoteValue(for flag: FeatureFlag) {
```

Which would result in calls like this one:

```swift
remoteFeatureFlagStore.hasRemoteValue(for: .newLoginFlow)
```

File: WordPress/Classes/Stores/RemoteFeatureFlagStore.swift
Comment (diegoreymendez) [inline WordPress/Classes/Stores/RemoteFeatureFlagStore.swift]: Same comment as above:

```swift
public func value(for flag: FeatureFlag) -> Bool {
```

Which would result in calls like this one:

```swift
remoteFeatureFlagStore.value(for: .newLoginFlow)
```

File: WordPress/Classes/Stores/RemoteFeatureFlagStore.swift
Comment (diegoreymendez) [inline WordPress/Classes/Stores/RemoteFeatureFlagStore.swift]: IMHO this comment here should be part of the documentation in `FeatureFlag.remoteKey` instead of being present here, since that's intrinsic to `remoteKey` and not really to this particular guard.

File: WordPress/Classes/Stores/RemoteFeatureFlagStore.swift
Comment (diegoreymendez) [inline WordPress/Classes/Stores/RemoteFeatureFlagStore.swift:24]: Love that we're using the `Result` type more!

**Review by diegoreymendez — CHANGES_REQUESTED** [review]
The comments I left within the code are minor and pretty much subjective.  They're for your consideration.

The only thing I'd really add to this PR, which I feel is important, is unit tests.  The new class you created looks great, and it's written in a way that unit testing it is just a step away.  Even the concurrency working fine could be tested to make sure all is good.

But overall this is a very nice PR.

File: WordPress/Classes/Stores/RemoteFeatureFlagStore.swift
Comment (mindgraffiti) [inline WordPress/Classes/Stores/RemoteFeatureFlagStore.swift]: nitpick. Let's go more swift-y with [comments](https://github.com/raywenderlich/swift-style-guide/#comments) instead of C-based

```suggestion
    /// Fetches remote feature flags from the server.
    ///
    /// - Parameter callback: An optional callback that can be used to update UI following the fetch. It is not called on the UI thread.
```

File: WordPress/Classes/Stores/RemoteFeatureFlagStore.swift
Comment (mindgraffiti) [inline WordPress/Classes/Stores/RemoteFeatureFlagStore.swift]: Let's deal with optionals early instead of force unwrapping

```suggestion
        guard let cachedKey = flag.remoteKey else {
            return false
        }

        return flag.remoteKey != nil && cache[cachedKey] != nil
```

File: WordPress/Classes/Stores/RemoteFeatureFlagStore.swift
Comment (jkmassel) [reply WordPress/Classes/Stores/RemoteFeatureFlagStore.swift]: Addressed in e481fa72713c06912e97e7274baf60300992f72d. I didn't use the commit suggestion just because I wanted to make the line break and indentation consistent with the other documented method in this class.

File: WordPress/Classes/Stores/RemoteFeatureFlagStore.swift
Comment (jkmassel) [reply WordPress/Classes/Stores/RemoteFeatureFlagStore.swift]: Addressed in c2015f3f1a05446e730105371d6f9b9d339ea7ba. I didn't use the commit suggestion because if we're using a `guard` statement we don't need the `flag.remoteKey != nil` check again in the `return` statement, so might as well get rid of it 🤷‍♂️

File: WordPress/Classes/Stores/RemoteFeatureFlagStore.swift
Comment (jkmassel) [reply WordPress/Classes/Stores/RemoteFeatureFlagStore.swift]: Addressed in 720d35f707545fff468ef82df7c0990ce22d8efa. I also changed the method name slightly – I realized I wasn't happy with one being `hasRemoteValue` and the other being `getValue`, so now it's `hasValue` and `getValue` for consistency.

File: WordPress/Classes/Stores/RemoteFeatureFlagStore.swift
Comment (jkmassel) [reply WordPress/Classes/Stores/RemoteFeatureFlagStore.swift]: Same as above!

**Comment by jkmassel** [comment]
@diegoreymendez – testing is included in 155f49229289527ec160f19492a03a6a72c5dc53. I had to rework `FeatureFlag` protocol conformance in order to make it possible to substitute a `MockFeaturedFlag` in `RemoteFeatureFlagStore` methods.

**Comment by diegoreymendez** [comment]
> I had to rework FeatureFlag protocol conformance in order to make it possible to substitute a MockFeaturedFlag in RemoteFeatureFlagStore methods.

Apologies for the extra work.  Thanks a ton for taking the time to do this.

File: WordPress/Classes/Stores/RemoteFeatureFlagStore.swift
Comment (jkmassel) [reply WordPress/Classes/Stores/RemoteFeatureFlagStore.swift]: I ended up putting it in both places in dd84a9bd8702897382058007f94d4ace1c393c8b. 

I agree that it's important to have it in the protocol definition for sure, but this _should_ be the only place in the codebase that this value is examined, and it may not be immediately clear _why_ it needs to be unwrapped when reading the code, which is why I left the comment in there.

Hope that's ok?

File: WordPress/WordPressTest/FeatureFlagTests.swift
Comment (diegoreymendez) [inline WordPress/WordPressTest/FeatureFlagTests.swift:4]: I believe that if you make this conform to `CaseIterable` you should be able to get all cases by using `MockFeatureFlag.allCases` (and not require the `remoteCases` array below).

Ref: https://www.simpleswiftguide.com/how-to-iterate-over-enum-cases-in-swift/

File: WordPress/WordPressTest/RemoteFeatureFlagTests.swift
Comment (diegoreymendez) [inline WordPress/WordPressTest/RemoteFeatureFlagTests.swift:18]: Nitpicky: I'd consider adding an expectation so that we know the `deviceIDCallback` is called twice.  The reason is that if the callback is never executed this test will never fail.

Something like:

```swift
func testThatDeviceIdIsTheSameForEveryInstanceOfTheStore() {
    let callbackExecuted = expectation(description: "Callback executed twice")
    callbackExecuted. expectedFulfillmentCount = 2

    let mock = MockFeatureFlagRemote()
    var deviceId = ""
    mock.deviceIdCallback = {
        deviceId == "" ? deviceId = $0 : XCTAssertEqual(deviceId, $0)
    }

    RemoteFeatureFlagStore(remote: mock).update()
    RemoteFeatureFlagStore(remote: mock).update()

    waitForExpectations(timeout: 0.1)
}
```

**Review by diegoreymendez — APPROVED** [review]
@jkmassel - Some optional comments, but I'm approving since this works as described.

Really nice work!

File: WordPress/WordPressTest/FeatureFlagTests.swift
Comment (jkmassel) [reply WordPress/WordPressTest/FeatureFlagTests.swift:4]: Yeah! I don't actually want all the cases – just the remote ones (so not `enabledFeature`, `disabledFeature`, and `nonOverrideableFeature` – I thought about using `CaseIterable` and filtering out the ones I _didn't_ want but thought maybe that'd be too clever? 🤷‍♂️

This way is a bit more explicit, if verbose 😅

File: WordPress/WordPressTest/RemoteFeatureFlagTests.swift
Comment (jkmassel) [reply WordPress/WordPressTest/RemoteFeatureFlagTests.swift:18]: This seems like a good extra check, so added in 7ba7d0946ec64e0e86576eb0cfc92854203cc07d :)

File: WordPress/WordPressTest/FeatureFlagTests.swift
Comment (diegoreymendez) [reply WordPress/WordPressTest/FeatureFlagTests.swift:4]: It's fine I didn't realize it was some.

**Review by diegoreymendez — APPROVED** [review]
I just retested this with the new WPKit version and it's working as described.

**Pull request merged**
